hop/subscribe.py

```diff
@@ -6,28 +6,61 @@
 
 import argparse
 import json
+import warnings
 
 from . import cli
 from .io import Stream
-from .models import GCNCircular
+from .models import GCNCircular, VOEvent
 
 
-def print_gcn(gcn_dict, json_dump=False):
-    """Parse a gcn dictionary and print to stdout.
+def print_msg(msg, json_dump=False):
+    """Check the format of a message obtained from an ADC stream, use it to instantiate
+    a data model corresponding to that format, and print the message.
 
     Args:
-      gcn_dict:  the dictionary object containing a GCN-formatted message
+      msg:       the raw message from an ADC stream
+      json_dump: boolean indicating whether to print msg as raw json
 
     Returns:
       None
-
     """
+
+    # check for msg format using standard-specific flags
```

Comment:
Oh man, this looks hard to maintain and destined to cause bugs: we're doing a string substring match to tell what type of message it is. This will take `O(n*m)` for `n` message formats of `m` length, which we will certainly regret eventually. Can we do something better? @adambrazier this has important architectural consequences: how do we discriminate between message types?

We should know based on the Kafka topic, but this code doesn't know what topic we're listening to at all - that's all abstracted away by `adc-streaming`, since we want to be agnostic to Kafka. For the same reason, we also can't use Kafka message metadata.

As I see it, that only leaves making a wrapper schema around any messages. Something like `{"type": "GCN Circular", "payload": {....}}`. We'd have to enforce that schema on the publish side probably. There are still choices to make: is JSON the right wrapper language, for example? The modern "right" answer in Kafka is almost certainly Avro.

It's fair to table this for after the workshop, but we really need to figure this out I think.

Reply:
I agree; I added it as a purely temporary placeholder since I'm certain there's a better way to do this with more time and by clarifying how we want to handle message type (at both the consumer and producer levels).

You're right -- the ultimate solution shouldn't rely on kafka topic, let alone message strings. There's more to think about for the right implementation (time for a user story), though for this release the substring holdover works for the AWS kafka instance.

Reply:
> It's fair to table this for after the workshop, but we really need to figure this out I think.

Definitely. If we have guarantees for what messages come out of certain topics, then we don't need message metadata to determine the message type. What we're lacking at the moment is a specification for our API/schema/serialization, however minimal that may be to be able to answer questions like these.